Detectors/ITSMFT/ITS/reconstruction/include/ITSReconstruction/Clusterer.h

```diff
@@ -0,0 +1,54 @@
+/// \file Clusterer.h
+/// \brief Definition of the ITS cluster finder
+#ifndef ALICEO2_ITS_CLUSTERER_H
+#define ALICEO2_ITS_CLUSTERER_H
+
+#include <utility>
+#include <vector>
+
+#include "Rtypes.h"
+
+class TClonesArray;
+
+namespace o2
+{
+namespace ITS
+{
+  class PixelReader;
+  class Clusterer
+{
+ public:
+  Clusterer();
+  ~Clusterer() = default;
+
+  Clusterer(const Clusterer&) = delete;
+  Clusterer& operator=(const Clusterer&) = delete;
+
+  static void setPixelGeometry(Float_t px, Float_t pz, Float_t x0, Float_t z0) {
+    mPitchX=px; mPitchZ=pz; mX0=x0; mZ0=z0;
+  }
+  void process(PixelReader &r, TClonesArray &clusters);
+
+ private:
+  enum {kMaxRow=650}; //Anything larger than the real number of rows (512 for ALPIDE)
+  void initChip(UShort_t chipID, UShort_t row, UShort_t col, Int_t label);
+  void updateChip(UShort_t chipID, UShort_t row, UShort_t col, Int_t label);
+  void finishChip(TClonesArray &clusters);
+
+  Int_t mColumn1[kMaxRow+2];
+  Int_t mColumn2[kMaxRow+2];
+  Int_t *mCurr, *mPrev;
+  std::vector<std::vector<std::pair<UShort_t,UShort_t>>> mPreClusters;
```

Comment:
Ok, then if I understand correctly I would do the following:

```C++
struct Digit {
  int cluster_index;
  short row;
  short col;
};

std::vector<Digit> mPreClusters;
```

This way you will pay 4 bytes per digit rather than 24 bytes per cluster, with the added advantage they are not scattered in memory. This should be a win in terms of memory if the typical number of Digits per Cluster is <6. Filling the vector will be much faster (because there is no per Cluster allocation). If you sort the container (or you insert in a cluster-sorted way) sequential access will be much faster (because you do not jump around in memory) and even random access should be fast enough given you can do binary search.